File: in-one-weekend/src/hittable_list.rs
use crate::{HitRecord, Hittable};
use common::ray::Ray;
use std::cell::RefCell;
use std::rc::Rc;

pub struct HittableList {
    pub objects: Vec<Rc<RefCell<dyn Hittable>>>,
}

impl HittableList {
    pub fn new() -> Self
    where
        Self: Sized,
    {
        HittableList {
            objects: Vec::new(),
        }
    }

    pub fn clear(&mut self) {
        self.objects.clear();
    }

    pub fn add(&mut self, object: Rc<RefCell<dyn Hittable>>) {
        self.objects.push(object);
    }
}

impl Hittable for HittableList {
    fn hit(&self, ray: &Ray, t_min: f64, t_max: f64, rec: &mut HitRecord) -> bool {
        let mut temp_rec = HitRecord::new();
        let mut hit_anything = false;
        let mut closest_so_far = t_max;

        for object in &self.objects {
            if object
                .borrow()
                .hit(ray, t_min, closest_so_far, &mut temp_rec)
            {
                hit_anything = true;
                closest_so_far = temp_rec.t;
                *rec = temp_rec.clone();
            }
        }

        hit_anything
    }
}

File: tj/hittable_list.py
import time

import taichi as ti
from taichi.math import *

from ray import Ray
from hittable import Hittable, HitRecord, Shape
from sphere import Sphere


class HittableList(Hittable):
    def __init__(self):
        super(HittableList, self).__init__(1)
        self.OS = None
        self.objs = []
        self.objs_len = 0

    def append(self, o: ti.template()):
        self.objs.append(o)

    def make_tj(self):
        self.objs_len = len(self.objs)
        OS = Sphere.field()
        ti.root.dense(ti.i, self.objs_len).place(OS)
        for i in range(OS.shape[0]):
            OS[i] = self.objs[i]
        self.OS = OS

    @ti.func
    def hit(self, ray: Ray, t_min: float, t_max: float) -> (bool, HitRecord):
        hit_anything = False
        closest_so_far = t_max
        rec = HitRecord()
        for i in range(self.objs_len):
            O = self.OS[i]
            hit, temp_rec = O.hit(ray, t_min, closest_so_far)
            if hit:
                hit_anything = True
                closest_so_far = temp_rec.t
                rec = temp_rec
        return hit_anything, rec
